Assets/IATK/Scripts/Model/RealtimeDataSource.cs

```diff
@@ -0,0 +1,384 @@
+        /// <returns>True if successfully added, false otherwise</returns>
+        public bool AddDimension(string dimensionName, float minVal, float maxVal, DataType type = DataType.Float)
+        {
+            // Don't add the dimension if it already exists
+            if (textualDimensionsList.ContainsKey(dimensionName)) return false;
+
+            var metaData = new DimensionData.Metadata();
+            metaData.minValue = minVal;
+            metaData.maxValue = maxVal;
+            metaData.type = type;
+
+            int index = dimensionData.Count;
+
+            textualDimensionsList.Add(dimensionName, new Dictionary<int, string>());
+            textualDimensionsListReverse.Add(dimensionName, new Dictionary<string, int>());
+
+            var dd = new DimensionData(dimensionName, index, metaData);
+            dd.setData(GetDefaultArray(), textualDimensionsList);
+            dimensionData.Add(dd);
+            dimensionPointers.Add(0);
+
+            //Debug.Log("RTDS AddDimension => " + dd.Identifier + ", " + dd.Index);
+            return true;
+        }
+
+        /// <summary>
+        /// Sets a data value by index
+        /// </summary>
+        /// <param name="index">Index of dimension</param>
+        /// <param name="val">Value to set the dimension</param>
+        /// <returns>True if successfully set, false otherwise</returns>
+        public bool SetData(int index, float val)
+        {
+            return SetData(this[index].Identifier, val);
+        }
+
+
+        /// <summary>
+        /// This is important otherwise the overloading can NOT be resolved from outside (eg JS)
+        /// </summary>
+        /// <param name="dimensionName"></param>
+        /// <param name="val"></param>
+        /// <returns></returns>
+        public bool SetDataStrVal(string dimensionName, float val)
+        {
+            //Debug.Log("RTDS SetDataStrVal => " + dimensionName + ", " + val);
+            return SetData(dimensionName, val);
+        }
+
+        /// <summary>
+        /// This is important otherwise the overloading can NOT be resolved from outside (eg JS)
+        /// </summary>
+        /// <param name="dimensionName"></param>
+        /// <param name="val"></param>
+        /// <returns></returns>
+        public bool SetDataStrStr(string dimensionName, string val)
+        {
+            return SetData(dimensionName, val);
+        }
+        /// <summary>
+        /// Sets a data value by dimension name (identifier)
+        /// </summary>
+        /// <param name="index">Name of dimension</param>
+        /// <param name="val">Value to set the dimension</param>
+        /// <returns>True if successfully set, false otherwise</returns>
+        public bool SetData(string dimensionName, float val)
+        {
+            try
+            {
+                var dd = this[dimensionName];
+
+                //this is needed since we do not know the data
+                //auto scale start
```

Comment:
Yes, I understand, but perhaps we can handle this separately in a different function? It would be nice if we could have one function that restricts the incoming data to be within the min and max, and then have a second function that auto scales the data?

Let me know your thoughts 🙂

Reply:
Yes I think the auto-scaling way you chose is a good solution for datasets with unknown min and max values.

When dealing with a static dataset (such as a CSV) it is easy to tell the min and max value, but when it comes to realtime data there might be no way of knowing the range, so I think having two versions might be best, one with the auto-scaling and then one without 🙂